pkg/controllers/securitypolicy_controller.go

```diff
@@ -17,37 +21,106 @@ import (
 )
 
 var (
-	log = logf.Log.WithName("controller").WithName("securitypolicy")
+	log           = logf.Log.WithName("controller").WithName("securitypolicy")
+	GCInterval    = 60 * time.Second
+	finalizerName = "securitypolicy.nsx.vmware.com/finalizer"
 )
 
 // SecurityPolicyReconciler reconciles a SecurityPolicy object
 type SecurityPolicyReconciler struct {
-	client.Client
+	Client  client.Client
 	Scheme  *runtime.Scheme
 	Service *services.SecurityPolicyService
 }
 
 func (r *SecurityPolicyReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-
 	obj := &v1alpha1.SecurityPolicy{}
-	if err := r.Get(ctx, req.NamespacedName, obj); err != nil {
-		log.Error(err, "unable to fetch object")
-		return ctrl.Result{}, err
+	log.Info("reconciling securitypolicy CR", "securitypolicy", req.NamespacedName)
+
+	if err := r.Client.Get(ctx, req.NamespacedName, obj); err != nil {
+		log.Error(err, "unable to fetch security policy CR")
+		return ctrl.Result{}, client.IgnoreNotFound(err)
 	}
 
-	log.Info("reconciling securitypolicy CR", "securitypolicy", req.NamespacedName)
+	if obj.ObjectMeta.DeletionTimestamp.IsZero() {
+		if err := r.Service.CreateOrUpdateSecurityPolicy(obj); err != nil {
+			log.Error(err, "failed to create or update security policy", "securitypolicy", req.NamespacedName)
+			return ctrl.Result{}, err
+		}
+		if !controllerutil.ContainsFinalizer(obj, finalizerName) {
+			controllerutil.AddFinalizer(obj, finalizerName)
+			if err := r.Client.Update(ctx, obj); err != nil {
+				return ctrl.Result{}, err
+			}
+		}
```

Comment:
Just in case, I think we'd better add the finalizer before `CreateOrUpdateSecurityPolicy`. Because I think the finalizer is a signal that the operator takes the CR into its management scope.  I'm not sure if there's a corner case that the CR is deleted after nsx policy being created but before adding finalizer.

Reply:
agree